MM-22896: Update profile popover UI

sinansonmez opened the pull request:
#### Summary
This PR updates profile popover UI

#### Ticket Link
Fixes https://github.com/mattermost/mattermost-server/issues/21192
JIRA: https://mattermost.atlassian.net/browse/MM-22896

#### Screenshots
**Other User**
|  before  |  after  |
|----|----|
| <img width="266" alt="Screen Shot 2022-11-06 at 22 09 01" src="https://user-images.githubusercontent.com/37421564/200195378-bd105f25-9ee8-4b13-a8f8-35034b494a90.png"> | <img width="224" alt="Screen Shot 2022-11-06 at 22 07 37" src="https://user-images.githubusercontent.com/37421564/200195409-99b26006-d7e5-4ecb-9c25-fb3a64788a9d.png"> |


**Current User**
|  before  |  after  |
|----|----|
| <img width="277" alt="Screen Shot 2022-11-06 at 22 08 52" src="https://user-images.githubusercontent.com/37421564/200195452-99ad124c-ae5f-46be-884e-a89a4c089e68.png"> | <img width="210" alt="Screen Shot 2022-11-06 at 22 07 49" src="https://user-images.githubusercontent.com/37421564/200195479-ccc0aa84-e2c8-475c-ad62-a5e08ef80896.png"> |


#### Release Note
```release-note
Update profile popover
```


[MM-22896]: https://mattermost.atlassian.net/browse/MM-22896?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

mattermod commented:
Hello @sinansonmez,

Thanks for your pull request! A Core Committer will review your pull request soon. For code contributions, you can learn more about the review process [here](https://developers.mattermost.com/contribute/getting-started/code-review/).

mattermod commented:
E2E tests not automatically triggered, because PR has no approval yet. Please ask a developer to review and then try again to attach the QA label.

sinansonmez commented:
@nevyangelova could you please run e2e tests. I couldnt run them locally. Probably I need to update some of them

sinansonmez commented on plugins/call_button/call_button.tsx:
```diff
@@ -44,25 +45,17 @@ export default function CallButton({pluginCallComponents, currentChannel, channe
                 onTouchEnd={clickEnabled ? clickHandler : undefined}
             >
-                {item.button}
+                {customButton || item.button}
```
This part is updated to be able to pass custom bottom. However I believe part below should be also updated but I couldnt test this (i.e.  when `pluginCallComponents.length` is bigger than 1) locally. Please guide me how to test this locally so I can check and update the UI if needed

https://github.com/mattermost/mattermost-webapp/blob/e0d323001bb151be0d6eb539ce96cc989565a522/plugins/call_button/call_button.tsx#L73

sinansonmez commented:
/update-branch

sinansonmez commented:
/e2e-test

mattermod commented:
You don't have permissions to trigger this command.
 It's only available for organization members.

cyrilzhang-mm commented on components/profile_popover/profile_popover.tsx:
```diff
@@ -151,19 +158,19 @@ interface ProfilePopoverProps extends Omit<React.ComponentProps<typeof Popover>,
 }
 type ProfilePopoverState = {
     loadingDMChannel?: string;
+    show: boolean;
```
The visibility of the profile popover isn't really handled by the popover itself, it's handled by whatever opened it. Using this `show` variable to hide the popover will lead to conflicts with the state inside and outside the popover. Try closing the popover with the X button and then opening it again; it results in a JavaScript error. 

Instead, to close the popover, you can simply call `this.props.hide()`. So I would do that in `handleClose` and remove the `show` variable.

sinansonmez replied:
Good point @cyrilzhang-mm I update accordingly

nevyangelova commented:
/e2e-test

mattermod commented:
Successfully triggered e2e testing!
https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/259299

matthewbirtch reviewed (commented):
Thanks for working on this one @sinansonmez. I'm seeing quite a few inconsistencies with the design and given this is largely a UI update, it's important to work out the small details. 

This one was a little harder to comment directly on the CSS, so I'm going to have to describe the updates needed with text and screenshots. I'll try to go from top to bottom. Let me know if you need clarity on any of the points below.

For starters, a comparison of design to implementation:

<table>
<tr></tr>
<th>Design</th>
<th>Implemented</th>
</tr>
<tr></tr>
<td>
<img src="https://user-images.githubusercontent.com/2040554/200344923-26905de2-3fc7-4eb9-99f8-ed57be80bacb.png" width="240">
</td>
<td>
<img src="https://user-images.githubusercontent.com/2040554/200347671-87e153e2-3913-4ec9-a918-e3ea6b56b413.png" width="185">
</td>
</tr>
</table>

1. The width of the popover looks to be a variable width. Can we give it a min-width of 200px and a max-width of 240px?
2. The user role tag (that shows for system admins) is getting the wrong font. I think because the container it is located in is an `<h3>` tag for some reason. Also, the container is called `popover-title` which isn't quite accurate since this container houses the close button and the user role tag.
3. For the user role tag, I believe there is a `tag` component that we should use for this.
4. For the close `x` button, it isn't positioned quite right. It looks like it needs to be pushed 4px closer to the right edge
5. For the avatar, I'm actually thinking now that I'm playing with it in a real context, we can go bigger. Can we update to 120x120? I'll make the change in the design file as well.
6. The status icon that overlaps the avatar should be updated to use the compass icon `icon-check-circle`. Once you've done this, you may need to adjust the size and position to match the design as well.
7. Can we add 8px of margin below the user profile image?
8. The font used for the display name (or username if there is no display name) should be using `font-family: Metropolis,sans-serif;`
9. It looks like we have the display name getting truncated if it gets too long and is set to `no-wrap`. I don't think we should do that in this case since the point of the popover is to provide more details on the user. We should allow the name to wrap on 2 lines if needed.
10. There should be a divider line that divides the top centered content from the bottom left-aligned content (see design above). Once you've added that, you may need to adjust padding or margins to match spacing in the design
11. For the sections in the bottom, we need to add 4px of margin between the title (e.g. Local Time) and the value (e.g. 10:26 AM EST). 
12. Between each of these sections, the space between them needs to be 8px
13. When custom status is set, can we add 4px of margin to the right of the emoji (if there is one)?
14. For the footer, can we make the padding 16px on all sides of the footer container?
15. The primary button on the left needs a hover state to match the standard primary button
16. The primary button is supposed to be 32px in height, but it looks to be 30.
17. The icon that's used in the primary button looks a bit smaller to me and is using an SVG rather than a compass icon. Can we switch this to use the compass icon?
18. For the icon buttons that are on the right side of the footer, their containers should be rounded squares, but they look wider than they are tall. They should be 32x32
19. The 'Start call' button should only show if it's available in the DM. In the system console the setting `Enable on all channels` was `false`, but the call button still showed on the profile popover. We probably need to do a check if the DM with that user has calls enabled and then hide it if it's not enabled
20. I forgot to mention that the start call should be disabled if you're already on a call with the person.

sinansonmez commented:
/update-branchs

mattermod commented:
Error trying to update the PR.
Please do it manually.

sinansonmez commented on components/profile_popover/profile_popover.tsx:
```diff
@@ -617,31 +619,50 @@ class ProfilePopover extends React.PureComponent<ProfilePopoverProps, ProfilePop
+                id: t('webapp.mattermost.feature.start_call'),
+                defaultMessage: 'Start Call',
+            });
+            const iconButtonClassname = classNames('btn icon-btn', {'icon-btn-disabled': disabled});
```
I couldnt use `disabled` prop for button. Because when I use, for some reason tooltip is not shown when the button is disabled

sinansonmez replied:
Thanks `aria-disabled` is added. `aria-label` is already added to `PhoneInTalkIcon` component. Do you think it is still necessary add `aria-label` again to wrapping component (i.e. button with `id='startCallButton'`)

cyrilzhang-mm replied:
You are right, I think that's fine.

sinansonmez commented on components/profile_popover/profile_popover.tsx:
```diff
@@ -617,31 +619,50 @@ class ProfilePopover extends React.PureComponent<ProfilePopoverProps, ProfilePop
+            const disabled = this.props.isUserInCall || this.props.isCurrentUserInCall;
+            const startCallMessage = this.props.isUserInCall ? formatMessage({
+                id: t('user_profile.call.userBusy'),
+                defaultMessage: 'User is in another call',
```
@matthewbirtch this message is shown to user in tooltip, when the user that current user trying to call is in another call. Please feel free to suggest different wording.

matthewbirtch replied:
Nice! I like this

sinansonmez replied:
Thanks, done

sinansonmez commented on packages/mattermost-redux/src/selectors/entities/common.ts:
```diff
@@ -51,3 +52,12 @@ export function getCurrentUserId(state: GlobalState): string {
+
+// Calls
+
+export function getCalls(state: GlobalState): Record<string, UserProfile[]> {
```
I added calls related selectors in this file. Feel free to suggest any changes

sinansonmez commented on packages/mattermost-redux/src/selectors/entities/common.ts:
```diff
@@ -51,3 +52,12 @@ export function getCurrentUserId(state: GlobalState): string {
+export function getCalls(state: GlobalState): Record<string, UserProfile[]> {
+    const callsPlugin = 'plugins-' + suitePluginIds.calls;
+
+    //@ts-ignore
```
I had to add `//@ts-ignore` as `GlobalState` currently doesnt have `plugins-com.mattermost.calls` typed. Please let me know can I avoid this

sinansonmez commented:
/update-branch

sinansonmez commented:
@matthewbirtch I still need to update e2e tests, but this is ready for your 2nd round of review from UX perspective. I had a couple of questions below related to some of your comments. The rest is addressed. 


> 2. The user role tag (that shows for system admins) is getting the wrong font. I think because the container it is located in is an h3 tag for some reason. Also, the container is called popover-title which isn't quite accurate since this container houses the close button and the user role tag.

Usage of h3 tag and ‘popover-title’ class is coming from third party library (react-bootstrap) so I couldnt change it

> 6. The status icon that overlaps the avatar should be updated to use the compass icon icon-check-circle. Once you've done this, you may need to adjust the size and position to match the design as well.

In order to change status avatar, all the files below for all statuses should be changed. I believe this is out of scope of this PR
import StatusAwayAvatarIcon from 'components/widgets/icons/status_away_avatar_icon';
import StatusAwayIcon from 'components/widgets/icons/status_away_icon';
import StatusDndAvatarIcon from 'components/widgets/icons/status_dnd_avatar_icon';
import StatusDndIcon from 'components/widgets/icons/status_dnd_icon';
import StatusOfflineAvatarIcon from 'components/widgets/icons/status_offline_avatar_icon';
import StatusOfflineIcon from 'components/widgets/icons/status_offline_icon';
import StatusOnlineAvatarIcon from 'components/widgets/icons/status_online_avatar_icon';
import StatusOnlineIcon from 'components/widgets/icons/status_online_icon';

> 9. It looks like we have the display name getting truncated if it gets too long and is set to no-wrap. I don't think we should do that in this case since the point of the popover is to provide more details on the user. We should allow the name to wrap on 2 lines if needed.

Should we also do it for role and username? Currently, role and username as well are truncated if it gets too long.

> 13. When custom status is set, can we add 4px of margin to the right of the emoji (if there is one)?

AFAIK, It is not possible to set status without emoji. Currently margin is always set after emoji. 

> 15. The primary button on the left needs a hover state to match the standard primary button

What should be the hover state? I couldnt see the state in the figma file.

jgilliam17 commented:
/e2e-test

mattermod commented:
Successfully triggered e2e testing!
https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/262014

matthewbirtch commented:
> > 3\. The user role tag (that shows for system admins) is getting the wrong font. I think because the container it is located in is an h3 tag for some reason. Also, the container is called popover-title which isn't quite accurate since this container houses the close button and the user role tag.
> 
> Usage of h3 tag and ‘popover-title’ class is coming from third party library (react-bootstrap) so I couldnt change it

Actually, it looks like it's using the proper `Badge` component now and is no longer using the h3 tag

> > 6\. The status icon that overlaps the avatar should be updated to use the compass icon icon-check-circle. Once you've done this, you may need to adjust the size and position to match the design as well.
> 
> In order to change status avatar, all the files below for all statuses should be changed. I believe this is out of scope of this PR import StatusAwayAvatarIcon from 'components/widgets/icons/status_away_avatar_icon'; import StatusAwayIcon from 'components/widgets/icons/status_away_icon'; import StatusDndAvatarIcon from 'components/widgets/icons/status_dnd_avatar_icon'; import StatusDndIcon from 'components/widgets/icons/status_dnd_icon'; import StatusOfflineAvatarIcon from 'components/widgets/icons/status_offline_avatar_icon'; import StatusOfflineIcon from 'components/widgets/icons/status_offline_icon'; import StatusOnlineAvatarIcon from 'components/widgets/icons/status_online_avatar_icon'; import StatusOnlineIcon from 'components/widgets/icons/status_online_icon';

Yep, that's fine. This can be a separate PR. I'll create a ticket for that

> > 9\. It looks like we have the display name getting truncated if it gets too long and is set to no-wrap. I don't think we should do that in this case since the point of the popover is to provide more details on the user. We should allow the name to wrap on 2 lines if needed.
> 
> Should we also do it for role and username? Currently, role and username as well are truncated if it gets too long.

Role will definitely need to remain truncated, but I think we could remove the truncation from username

> > 29\. The primary button on the left needs a hover state to match the standard primary button
> 
> What should be the hover state? I couldnt see the state in the figma file.

This is what we use for the hover state on the primary button elsewhere (you can see it on the Create a new channel modal, for example):
```
background: linear-gradient(0deg, rgba(0, 0, 0, 0.08), rgba(0, 0, 0, 0.08)),var(--button-bg);
```

> 19\. The 'Start call' button should only show if it's available in the DM. In the system console the setting `Enable on all channels` was `false`, but the call button still showed on the profile popover. We probably need to do a check if the DM with that user has calls enabled and then hide it if it's not enabled

This one looks to be outstanding still.

jgilliam17 commented:
Thanks @sinansonmez 
Here are e2e reports, all failures on PR test are related to profile popover. 
PR test: https://automation-dashboard.vercel.app/cycles/8dbf47c6-eed9-4429-b33e-22ed4c60ab0c
latest master report for comparison: https://automation-dashboard.vercel.app/cycles/ebfd08dc-c648-4be4-895b-340b3793b58b

sinansonmez commented:
@matthewbirtch thanks your remaining points are address except

sinansonmez commented:
> Role will definitely need to remain truncated, but I think we could remove the truncation from username

>This is what we use for the hover state on the primary button elsewhere (you can see it on the Create a new channel modal, for example):

@matthewbirtch 2 points are addressed. Related to the point below, I see the issue. In order to work it properly, you need to refresh the page. Please see the recording below. As I dont have detailed info how plugins are integrating with the webapp. Maybe @BenCookie95 can support me during dev review


https://user-images.githubusercontent.com/37421564/202019419-38d07247-d2df-470d-befa-27ae154e809f.mov

BenCookie95 commented:
> > Role will definitely need to remain truncated, but I think we could remove the truncation from username
> 
> > This is what we use for the hover state on the primary button elsewhere (you can see it on the Create a new channel modal, for example):
> 
> @matthewbirtch 2 points are addressed. Related to the point below, I see the issue. In order to work it properly, you need to refresh the page. Please see the recording below. As I dont have detailed info how plugins are integrating with the webapp. Maybe @BenCookie95 can support me during dev review
> 
>  Screen.Recording.2022-11-15.at.21.26.30.mov

@sinansonmez Can you see any websocket messages that happen when you disable calls from system console? It looks like the start call button is also still in the channel header before you refresh the page as well. We might need to file a ticket with the calls team to address that

sinansonmez commented:
no worries @matthewbirtch 

> two more last UI things:

I will check both of them

> If I'm in a DM with the same user I want to call with, but calls are disabled in the DM header menu (screenshot above), we should hide the calls button in this case as well.

This is actually implemented, but I have the same issue here. The page needs to be refreshed in order to this work. I will check based on Ben's comments. 

As I am thinking of `Enable on all channels` and `Enable on specific channels`, we need to be more structured to make sure that we are covering everything. In order to do this, I list all possible scenarios in the table below. Please check if I am missing anything. After your feedback, I will double check the code again based on this table. 

|  Enable on all channels | Enable on specific channels | Channel Setting (Enabled / Disabled)  |  Action (Show / Hide Start Call Button)  |
| ------------- | ------------- | ------------- | ------------- |
| True  | False  | Not applicable | Show Button |
| True  | True  | Enabled | Show Button |
| True  | True  | Disabled | Hide Button |
| False  | False | Not applicable | Hide Button |
| False  | True  | Enabled | Show Button |
| False  | True  | Disabled | Hide Button |

sinansonmez commented:
Thanks @BenCookie95 when I change the setting, I dont see any message in the system console. I only see config in the network tab which is expected to update the config. I will wait for Chris's feedback on this
<img width="544" alt="Screen Shot 2022-11-16 at 21 34 08" src="https://user-images.githubusercontent.com/37421564/202288560-76f84e8d-156a-407b-a725-c870736bfef9.png">

matthewbirtch commented:
> As I am thinking of `Enable on all channels` and `Enable on specific channels`, we need to be more structured to make sure that we are covering everything. In order to do this, I list all possible scenarios in the table below. Please check if I am missing anything. After your feedback, I will double check the code again based on this table.
> 

Enable on all channels | Enable on specific channels | Channel Setting (Enabled / Disabled) | Action (Show / Hide Start Call Button)
-- | -- | -- | --
True | False | Not applicable | Show Button
True | True | Enabled | Show Button
True | True | Disabled | Hide Button
False | False | Not applicable | Hide Button
False | True | Enabled | Show Button
False | True | Disabled | Hide Button

Table looks good to me @sinansonmez

cpoile commented:
Hey @sinansonmez, you get a `custom_com.mattermost.calls_channel_disable_voice` message for the channel where calls were disabled, and likewise a `...enable_voice` message. But you have to already be a member of that channel. If you're not a member, you won't get the websocket event.
<img width="685" alt="image" src="https://user-images.githubusercontent.com/1490756/203157563-dcd2f0d8-27c6-45ec-8fee-a98b6bd63ea3.png">

Let me know if that helps, or if you need more info.

cyrilzhang-mm commented:
Hey @sinansonmez I have just merged a PR (https://github.com/mattermost/mattermost-webapp/pull/11530) to improve keyboard accessibility in the profile popover, so you will probably have some merge conflicts. When you go to merge with master, here are some things to watch out for:
- All of the content should go in the popover body, i.e., don't pass `title` prop to `Popover`, just render the title inside the popover. Everything in the popover will be in a `div` with `role='dialog'` and `className={A11yClassNames.POPUP}` and the same `aria-label`, `onKeyDown` that I added. We put the `tabCatcher` element before and after this `div`. You will probably need to edit the CSS accordingly.
- In my changes, I have added the comment `// Focus the title when the popover first opens, to bring the focus into the popover.` In your changes, you will want to change this code to focus the close button that you added, instead of the title.

Let me know if you have any more problems with merging and I'll be happy to help. So sorry for the inconvenience :(

sinansonmez commented:
/update-branch

sinansonmez commented:
@cyrilzhang-mm thanks for the info. I believe I was able to properly merge the conflicts and address your 2 points mentioned below. 

> Hey @sinansonmez I have just merged a PR (#11530) to improve keyboard accessibility in the profile popover, so you will probably have some merge conflicts. When you go to merge with master, here are some things to watch out for:
> 
> * All of the content should go in the popover body, i.e., don't pass `title` prop to `Popover`, just render the title inside the popover. Everything in the popover will be in a `div` with `role='dialog'` and `className={A11yClassNames.POPUP}` and the same `aria-label`, `onKeyDown` that I added. We put the `tabCatcher` element before and after this `div`. You will probably need to edit the CSS accordingly.
> * In my changes, I have added the comment `// Focus the title when the popover first opens, to bring the focus into the popover.` In your changes, you will want to change this code to focus the close button that you added, instead of the title.
> 
> Let me know if you have any more problems with merging and I'll be happy to help. So sorry for the inconvenience :(

@cyrilzhang-mm thanks for the info. I believe I was able to properly merge the conflicts and address your 2 points mentioned above. Could you please take a look?

sinansonmez commented:
> Hey @sinansonmez, you get a `custom_com.mattermost.calls_channel_disable_voice` message for the channel where calls were disabled, and likewise a `...enable_voice` message. But you have to already be a member of that channel. If you're not a member, you won't get the websocket event. <img alt="image" width="685" src="https://user-images.githubusercontent.com/1490756/203157563-dcd2f0d8-27c6-45ec-8fee-a98b6bd63ea3.png">
> 
> Let me know if that helps, or if you need more info.

@cpoile thanks, I will check and come back. Sorry for late response

sinansonmez commented:
> Sorry @sinansonmez two more last UI things:
> 
> * If role tag does not show, can we reduce some of the white space above the avatar? I'd say we can reduce by 8px. Leave it as-is though if a role tag is visible
> * I just noticed the icons within the `add to channel` and `start call` buttons don't align properly with each other. It looks like the add to channel icon is sitting too high
> 
> <img alt="image" width="237" src="https://user-images.githubusercontent.com/2040554/202216429-6a68c4fd-4669-439a-bd3b-3ef52c1ed44f.png">


@matthewbirtch  2 points related to UI is addressed. Point related to hiding call button is WIP

sinansonmez commented on e2e/cypress/tests/integration/messaging/autocomplete_spec.js:
```diff
@@ -319,29 +319,4 @@ describe('autocomplete', () => {
         });
     });
-
-    it('MM-T2214 @ mention from link in profile popover: center', () => {
```
This part is deleted because in the new design there is no option to mention user when clicking the username

sinansonmez commented on packages/mattermost-redux/src/selectors/entities/common.ts:
```diff
@@ -8,6 +8,21 @@ import {TeamMembership} from '@mattermost/types/teams';
+
+const CALLS_PLUGIN = 'plugins-' + suitePluginIds.calls;
+
+// todo where to put this type. This type is already defined in calls plugin. How can I reuse in order to avoid duplication
```
if the location of the types are OK, I can remove the comment

jgilliam17 commented:
/e2e-test

mattermod commented:
Successfully triggered e2e testing!
https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/267232

sinansonmez commented:
The logic to show/hide call button updated based on the table above. 

Related to behavior when changing Call plugin settings (i.e. `Enable on all channels` and `Enable on specific channels`), I see that my implementation is same as current behavior of `Start Call` button. Please see the screen recording below. For this reason, I believe this PR is ready to proceed for reviews cc: @matthewbirtch @cpoile 

https://user-images.githubusercontent.com/37421564/204140627-59c6540e-bfb6-405e-b7a1-d0eeecce5e99.mov

sinansonmez commented on components/profile_popover/index.ts:
```diff
@@ -67,9 +89,15 @@ function makeMapStateToProps() {
         const enableLastActiveTime = displayLastActiveLabel(state, userId);
+        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+        // @ts-ignore
+        const isCallsEnabled = Boolean(state.plugins.plugins[suitePluginIds.calls]) && Boolean(state['plugins-' + suitePluginIds.calls]);
```
I had to add ignore comment as `plugins` is not very well typed in `GlobalState`

matthewbirtch reviewed (changes requested):
@sinansonmez I don't think this is still working correctly. 

@cpoile I'm hoping you can help out to make sure that the calls button is displaying appropriately on the popover when viewing within a DM — based on the system settings and the channel-specific settings. Chris is definitely more familiar with how calls settings are expected to work. It does feel a bit odd to me in the following case:

- In the system console, I turned `Enable on all channels` to `true`
- I want back to the chat-facing side of things
- Had to refresh in order for any change to take effect
- the `Start call` button in the channel header **does not** display 
- the `Start call` button on the profile popover **does** display (it shouldn't in this case)
- then from the DM menu in the header, choose `Enable calls`
- then open the profile popover and the call button within the profile popover shows for a split second and then hides.
- then from the DM menu in the header, choose `Disable calls`
- then open the profile popover and the call button within the profile popover shows - even though I've disabled it

<img width="963" alt="image" src="https://user-images.githubusercontent.com/2040554/204154442-3f548acd-8c40-4887-8d98-e1663af00a50.png">

<img width="957" alt="image" src="https://user-images.githubusercontent.com/2040554/204154865-22fe0d50-1364-4c7e-8423-30ac5ee2eb28.png">

I feel we should only show the call button in the profile popover if it is also showing in the channel header. Something is causing this to get out of sync. I'm a little confused with the settings in the system console though. 

One other new UI thing I noticed now:
- the SYSTEM ADMIN badge now has extra space on the left. Can remove the left margin that is applied here in this case?

cyrilzhang-mm reviewed (changes requested):
Thanks for doing the merge. I commented on a few minor things. As well, I noticed two visual issues with the focus state. They are very minor but if you can fix these that would be great.

Prevent the focus from being cut off here:
<img width="261" alt="Screen Shot 2022-11-28 at 9 25 12 AM" src="https://user-images.githubusercontent.com/112951043/204302251-a7ba4176-3b9c-4266-9685-6127445ec31d.png">

Make the button only take up its width, not the whole width of the popover
<img width="261" alt="Screen Shot 2022-11-28 at 9 25 18 AM" src="https://user-images.githubusercontent.com/112951043/204302255-7ca77ede-e8ac-4607-b0fd-13e77c3737f7.png">

sinansonmez commented on components/profile_popover/profile_popover.tsx:
```diff
@@ -715,67 +782,77 @@ ProfilePopoverState
-                {title}
                 {roleTitle}
+                <button
+                    tabIndex={-1}
```
Good to know, thanks

sinansonmez commented on components/profile_popover/profile_popover.tsx:
```diff
@@ -374,22 +373,32 @@ ProfilePopoverState
 
         return {customStatusContent, expiryContent};
     }
+    handleClose = () => this.props.hide?.();
```
Added, thanks

sinansonmez commented on components/profile_popover/profile_popover.tsx:
```diff
@@ -633,75 +645,130 @@ ProfilePopoverState
+                    </Tooltip>
+                }
+            >
+                <a>
```
When I remove `a` tag, tooltip is not working anymore on the `button` tab. When I check other buttons with tooltip in the codebase, I see similar behaviour (i.e. button is always wrapped with div). `Channel Files` button as an example below. Please let me know if you have any solution to overcome this

<img width="1134" alt="Screen Shot 2022-12-03 at 22 07 26" src="https://user-images.githubusercontent.com/37421564/205462204-08c6d3a6-c879-4965-8d4d-06e88609a52c.png">

sinansonmez replied:
Good suggestion. I didnt know about `SimpleTooltip` component. However, it didnt work again. Therefore I replaced `a` tag with `div` tag

sinansonmez commented:
@matthewbirtch I fined tuned the logic a bit. I summarized my logic below again. 

Please be aware the call button in the profile popover is mainly driven by call setting in DM channel. if user is enabled the call in DM channel, even though call is disabled in channel, the button will be hidden. Please see the first video below

- if call plugin is disabled --> dont show the button
- if user disable the calls in DM --> dont show the button in DM and also on all channels
- if system admin set `Enable on all channels` to false and `Enable on specific channels` and calls disabled in channel --> dont show the button when profile popover opened in channel (not DM channel). However, If user enabled the call in DM channel, show the button when profile popover opened in channel. Please see the video below

cyrilzhang-mm commented:
@matthewbirtch @sinansonmez 

> After doing a bit more testing, I think we may need to make some adjustments to the positioning logic.
> I'm not sure what logic is being used here, so it's difficult to suggest changes.

The popovers are positioned by whatever opened the popover. For now it seems like at-mentions are causing the problem so I suggest that in `at_mention.tsx` we change `spaceRequiredForPopOver` to a bigger number, roughly the height of a profile popover, and that would fix it for popovers opened by at-mentions. But it won't be perfect because not every profile popover is the same height so there might still be some edge cases that can cause it to go offscreen - unless there is a maximum height that profile popovers can be?

> In addition, I'm hoping we can add some motion to the fade in animation as follows

I'm not sure but this might be a bit difficult as in some places the visibility and position of the popover is controlled entirely by the Overlay component from react-bootstrap, i.e. the popovers don't know if they are visible or not or what their position is.

matthewbirtch commented:
> The popovers are positioned by whatever opened the popover. For now it seems like at-mentions are causing the problem so I suggest that in `at_mention.tsx` we change `spaceRequiredForPopOver` to a bigger number, roughly the height of a profile popover, and that would fix it for popovers opened by at-mentions. But it won't be perfect because not every profile popover is the same height so there might still be some edge cases that can cause it to go offscreen - unless there is a maximum height that profile popovers can be?

Yes, the height of a profile popover is a variable height and definitely is not consistent for all users. 

I'm not sure if a max-height has been added, but we should definitely add that. I would say we want to limit the height of a profile popover a max height similar to how will be limiting modals. This is essentially: 

100vh(viewport height) - 240px

If the content height goes beyond this, the popover should scroll. @sinansonmez do you think you could add this in as well?

Good catch @cyrilzhang-mm 

> > In addition, I'm hoping we can add some motion to the fade in animation as follows
> 
> I'm not sure but this might be a bit difficult as in some places the visibility and position of the popover is controlled entirely by the Overlay component from react-bootstrap, i.e. the popovers don't know if they are visible or not or what their position is.

I feel like we've begun to employ this type of motion in tour points, so to me it seems like the same concept. Again, maybe @AshishDhama can help with guidance on how this was accomplished for tour points.

sinansonmez commented:
@matthewbirtch @cyrilzhang-mm I have set the max-height to `100vh(viewport height) - 240px` and make the profile popover scrollable. I also put `100vh(viewport height) - 240px` for `spaceRequiredForPopOver ` in `at_mention.tsx`. This solved the issue for profile popovers opened by clicking mention tag. 

Related to adding motions, if technically possible for the reason @cyrilzhang-mm highlighted, I would suggest doing it in a separate ticket not to block this one further.

matthewbirtch reviewed (approved):
I'm good with creating a separate ticket for the motion updates @sinansonmez. This is a great improvement as-is @sinansonmez. Thanks so much!

sinansonmez commented:
great thanks @matthewbirtch. This PR is ready for another round of dev review @BenCookie95 @cyrilzhang-mm

cyrilzhang-mm commented on components/profile_popover/profile_popover.scss:
```diff
@@ -1,27 +1,140 @@
-    margin: 10px 0 4px;
-    color: rgba(var(--center-channel-color-rgb), 0.64);
+    margin: 8px 0 4px;
+    color: rgba(var(--center-channel-text-rgb), 0.64);
```
`--center-channel-color` is the correct variable to use here

cyrilzhang-mm commented on sass/components/_popover.scss:
```diff
@@ -274,43 +275,34 @@
+    border: none;
+    margin-left: auto;
+    background: none;
+    color: rgba(var(--center-channel-text-rgb), 0.56);
```
`--center-channel-color-rgb`

sinansonmez replied:
thanks

cyrilzhang-mm commented on components/profile_popover/profile_popover.tsx:
```diff
@@ -796,11 +883,7 @@ ProfilePopoverState
                     onKeyDown={this.handleKeyDown}
```
Let's add `aria-modal={true}` as well

sinansonmez replied:
added, thanks

cyrilzhang-mm commented on components/profile_popover/profile_popover.tsx:
```diff
@@ -604,12 +605,14 @@ ProfilePopoverState
-                            title={{
+                        <AccountOutlineIcon
+                            size={16}
+                            aria-label={{
```
Need to use the localization function here, this is just an object

sinansonmez replied:
yep you are right 😅

cyrilzhang-mm reviewed (changes requested):
Looks good @sinansonmez, I just have some small things for you to fix!

esethna commented:
Adding Laney as this is suite users AOR

sinansonmez commented on components/profile_popover/profile_popover.scss:
```diff
@@ -1,27 +1,140 @@
-    margin: 10px 0 4px;
-    color: rgba(var(--center-channel-color-rgb), 0.64);
+    margin: 8px 0 4px;
+    color: rgba(var(--center-channel-text-rgb), 0.64);
```
Thanks, changed

sinansonmez commented on components/profile_popover/profile_popover.scss:
```diff
@@ -1,27 +1,140 @@
+            gap: 12px;
+
+            &.first {
+                border-top-color: rgba(var(--center-channel-color-rgb), 0.08);
```
it is already using `--center-channel-color` it is just rgb version of it

sinansonmez commented on components/profile_popover/profile_popover.scss:
```diff
@@ -1,27 +1,140 @@
+        .user-popover__role {
+            padding: 0 4px;
+            margin-left: 0;
+            background-color: rgba(var(--center-channel-text-rgb), 0.08) !important;
```
good catch, it is removed

kostaspt reviewed (approved):
Great work @sinansonmez!

sinansonmez commented:
@BenCookie95 @kostaspt thanks for the approval. Could you please request a QA review

furqanmlk reviewed (commented):
@sinansonmez can you please fix the conflict?

sinansonmez commented:
@furqanmlk conflict is fixed

sinansonmez commented:
@furqanmlk thanks. Could you please merge it? Or should I ask from @BenCookie95 / @kostaspt ?

sinansonmez commented:
/update-branch

matthewbirtch commented:
What's left on this one @sinansonmez. Does it just need a PM review from @laneycs?

sinansonmez commented:
@matthewbirtch 2 devs and 1 QA approvals are completed. I am not sure if @laneycs need to give a review. If not needed, this is ready to be merged

sinansonmez commented:
@laneycs thanks for the approval
@matthewbirtch  @kostaspt @BenCookie95 @furqanmlk this is ready to be merged

AshishDhama commented:
/update-branch

AshishDhama commented on sass/components/_popover.scss:
```diff
@@ -200,6 +193,7 @@
         }
 
         &.divider--expanded {
+            border-color: rgba(var(--center-channel-text-rgb), 0.08);
```
```suggestion
            border-color: rgba(var(--center-channel-color-rgb), 0.08);
```

AshishDhama replied:
@sinansonmez Could you please change this,

AshishDhama replied:
@sinansonmez you missed this one, could you please update this as well.

AshishDhama commented on components/profile_popover/profile_popover.scss:
```diff
@@ -1,27 +1,140 @@
+        .user-popover__role {
+            padding: 0 4px;
+            margin-left: 0;
+            background-color: rgba(var(--center-channel-text-rgb), 0.08);
```
```suggestion
            background-color: rgba(var(--center-channel-color-rgb), 0.08);
```

AshishDhama commented on components/profile_popover/profile_popover.scss:
```diff
@@ -1,27 +1,140 @@
+            padding: 0 4px;
+            margin-left: 0;
+            background-color: rgba(var(--center-channel-text-rgb), 0.08);
+            color: var(--center-channel-text-rgb);
```
```suggestion
            color: var(--center-channel-color-rgb);
```

sinansonmez commented:
@AshishDhama thanks for the feedback. suggested changes are done

AshishDhama commented:
@sinansonmez Could you check if all of your changes are there, I can still see the same color variables?

sinansonmez commented:
@AshishDhama I checked again all four of my changes are reflected so you shouldnt see old variables anymore
